src/type-generator.ts

```diff
@@ -122,80 +152,107 @@ export class TypeGenerator {
+      }
 
-    if (def.type === 'number' || def.type === 'integer') {
-      return 'number';
+      return { type: 'Date', toJson: x => `${x}?.toISOString()` };
     }
 
-    if (def.type === 'string') {
-      if (def.format === 'date-time') {
-        return 'Date';
+    // enums
+    if (def.enum && Array.isArray(def.enum) && def.enum.length > 0 && !def.enum.find(x => typeof(x) !== 'string')) {
+      if (def.type && def.type !== 'string') {
+        throw new Error('only "string" enums are supported');
       }
 
-      if (Array.isArray(def.enum) && def.enum.length > 0 && !def.enum.find(x => typeof(x) !== 'string')) {
-        return this.emitEnum(typeName, def, structFqn);
+      return this.emitEnum(typeName, def, structFqn);
+    }
+
+    // struct
+    if (def.properties) {
+      if (def.type && def.type !== 'object') {
+        throw new Error('for "properties", if "type" is specified it as to be an "object"');
       }
 
-      return 'string';
+      return this.emitStruct(typeName, def, structFqn);
     }
 
     // map
-    if (!def.properties && def.additionalProperties && typeof(def.additionalProperties) === 'object') {
-      return `{ [key: string]: ${this.typeForProperty(typeName, def.additionalProperties)} }`;
+    if (def.additionalProperties && typeof(def.additionalProperties) === 'object') {
+      if (def.type && def.type !== 'object') {
+        throw new Error('for "additionalProperties", if "type" is specified it as to be an "object"');
+      }
+
+      const et = this.typeForProperty(typeName, def.additionalProperties);
+      const toJson = (x: string) => `((${x}) === undefined) ? undefined : (Object.entries(${x}).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: ${et.toJson('i[1]') } }), {}))`;
+      return { type: `{ [key: string]: ${et.type} }`, toJson };
     }
 
-    // struct
-    if (def.properties) {
-      this.emitStruct(typeName, def, structFqn);
-      return typeName;
+    switch (def.type) {
+      case 'string':
+        return { type: 'string', toJson: x => x };
+
+      case 'number':
+      case 'integer':
+        return { type: 'number', toJson: x => x };
+
+      case 'boolean':
+        return { type: 'boolean', toJson: (x) => x };
+
+      case 'array': {
+        return this.emitArray(typeName, def);
+      }
     }
 
-    return 'any';
+    return { type: 'any', toJson: x => x };
   }
 
   /**
    * Registers a custom type and emits it. This will override any existing
    * definitions for this type name.
    *
    * @param typeName The name of the type emitted by this handler.
-   * @param emitter A function that will be called to emit the code.
+   * @param emitter A function that will be called to emit the code and returns
+   * information about the emitted type.
    */
-  public emitCustomType(typeName: string, emitter: (code: Code) => void) {
-    if (this.emittedTypes.has(typeName)) {
+  public emitCustomType(typeName: string, emitter: TypeEmitter | CodeEmitter) {
```

Comment:
Is the union here just to avoid a breaking change? I don't see `CodeEmitter` being passed anywhere internally

Reply:
Yes, it is (and we are wrapping it as a TypeEmitter inside)